### PR TITLE
Implement seeking in MSCDEX

### LINKED ISSUE #3508: Implement seeking in MSCDEX
**Time Warriors** does not play CD Audio music in the main menu and during gameplay (reported here https://github.com/dosbox-staging/dosbox-staging/issues/3501). Note that CD Audio music _does_ play during the intro sequence.

ripsaw8080 did a fix for the in-game and menu music issue via a TSR, but that doesn't work in Staging. The root cause, however, is that the seek function of the emulated MSCDEX driver is not implemented.

Quoting his description of the issue ([source](https://www.vogons.org/viewtopic.php?p=299150#p299150)):

> The audio tracks are not working because the game is confused by the lack of support for the cdrom driver seek function. Very few games use seek, and this game seems to use it only to convert MSF to HSG sector numbers. It first reads the start sector for a particular track in MSF, calls the seek function with that MSF sector, calls the get position function to get the HSG sector, and finally calls the play function with the HSG sector. Seems pointless to me, but maybe the game developers had some reason to prefer HSG.
>
> The attached workaround program simulates seeks by converting them to zero length plays -- nothing actually plays, but the current position is updated. Run the program in DOSBox before running the game and the audio tracks should work. It also fixes the intro movie in [Alpha Storm](http://www.mobygames.com/game/dos/alpha-storm).

So implementing seek should improve CD Audio support for at least **Time Warriors** and **Alpha Storm**.

I've checked a few DOSBox forks, including X, and none of them implements the seek function, so we'll need to figure this out from scratch.

From `dos_mscdex.cpp`:

image

Well, no wonder it doesn't work! 😂

### PR DESCRIPTION
**Pull request by weirddan455**
# Description

Some games will use the seek function on the CDROM and then expect the position to be updated accordingly.  As far as I can tell, the only behavior of a seek is that the audio should stop and the position should be updated.

I tried doing this at the driver level but seeking in this way doesn't make sense for any of our drivers.  It's meaningless for a .iso file.  SDL_CD doesn't have a seek function.  The Linux ioctl CDROMSEEK exists but it fails on my system (probably needs an old school drive with an analogue output).

Solution is just fake it basically.  Whenever audio is not playing, just report that the current location is `audioStart` which gets set on seek or pause/play.  Again, trying to get the current position of the laser from a driver doesn't make sense when audio is not playing.

Documentation was pretty sparse but I found these helpful:

https://makbit.com/articles/mscdex.txt - Shows how to read the parameters for the interrupt

https://www.vogons.org/viewtopic.php?p=115298#p115298 - Someone mentioned the audio stop behavior of seek.

## Related issues

Fixes #3508

# Manual testing

- Tested an English and a French version of Time Warrirors - Music playback is fixed
- Tested Alpha Sorm - Opening movie is fixed
- Tested these with `IMGMOUNT` and with external cdemu software (tests the ioctl linux path)
- Tried to burn to a real CD but my burning software failed me.  Resulting disc was static and could not be played back with a regular music player.

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: src/dos/dos_mscdex.cpp [hunk 1]
```diff
@@ -915,6 +920,18 @@ bool CMscdex::GetChannelControl(uint8_t subUnit, TCtrl& ctrl) {
...
+	if (subUnit>=numDrives) return false;
+	dinfo[subUnit].lastResult = cdrom[subUnit]->StopAudio();
+	if (dinfo[subUnit].lastResult) {
+		dinfo[subUnit].audioPlay = false;
+		dinfo[subUnit].audioPaused = false;
+		dinfo[subUnit].audioStart = sector;
+		dinfo[subUnit].audioEnd = 0;
+	}
+	return dinfo[subUnit].lastResult;
+}
```
Comment (johnnovak) [inline src/dos/dos_mscdex.cpp:933]: Please clang-reformat this function.

File: src/dos/dos_mscdex.cpp [hunk 2]
```diff
@@ -1137,6 +1154,7 @@ static Bitu MSCDEX_Strategy_Handler(void) {
 	return CBRET_NONE;
 }
 
+// MSCDEX interrupt documentation: https://makbit.com/articles/mscdex.txt
```
Comment (johnnovak) [inline src/dos/dos_mscdex.cpp:1157]: 👍🏻

File: src/dos/dos_mscdex.cpp [hunk 3]
```diff
@@ -1178,8 +1196,25 @@ static Bitu MSCDEX_Interrupt_Handler(void) {
...
+							msf.fr = (sector >> 0) & 0xFF;
+							sector = msf_to_frames(msf);
+							if (sector < REDBOOK_FRAME_PADDING) {
+								MSCDEX_LOG("MSCDEX: Seek: invalid position %d:%d:%d", msf.min, msf.sec, msf.fr);
+								sector = 0;
+							} else {
+								sector -= REDBOOK_FRAME_PADDING;
+							}
+						}
+						mscdex->Seek(subUnit, sector);
```
Comment (johnnovak) [inline src/dos/dos_mscdex.cpp:1215]: Sending this through clang-format yields comical results, so maybe just leave it as is, yeah 😄

**Review by johnnovak — APPROVED** [review]
Very nice enhancement @weirddan455 🥳 Yeah, we only need JustEnough(tm) seek emulation to get these rare games that (mis)use the seek function running correctly.

I assume that the chance of the seek implementation causing regressions with other titles that use CD Audio is very low. I also assume it works with OGG, MP3, FLAC, etc. audio tracks, but I haven't actually tested it.

**Comment by johnnovak** [comment]
@Python-Exoproject This will fix the CD Audio problem with Time Warriors natively, without the TSR. `CDSEEK` must _not_ be loaded as that will "cancel out" the fix somehow.

It also fixes the CD Audio music in the opening video of Alpha Storm.

**Comment by weirddan455** [comment]
>  `CDSEEK` must not be loaded as that will "cancel out" the fix.

I haven't tested that program since it was only developed as a workaround for a Dosbox bug but (just based on the description of what it does), the reason it doesn't work on Staging is because we're explicitly ignoring zero length plays, at least for images.  Would need testing on real hardware if this is correct or not.

https://github.com/dosbox-staging/dosbox-staging/blob/6c87b92dad66709a9f1c134a373472df8cd58217/src/dos/cdrom_image.cpp#L710-L718

>  I assume that the chance of the seek implementation causing regressions with other titles that use CD Audio is very low. I also assume it works with OGG, MP3, FLAC, etc. audio tracks, but I haven't actually tested it.
>
>  I think it's safe to backport this to improve game compatibility. What's your opinion?

It's probably safe to backport.  I can do a bit of regression testing.  Is it documented anywhere how the .ogg files need to be structured?  I never really use that feature but I seem to recall it needs a .cue file to be written out in a specific way?

**Comment by MasterO2** [comment]
> > `CDSEEK` must not be loaded as that will "cancel out" the fix.
> 
> I haven't tested that program since it was only developed as a workaround for a Dosbox bug but (just based on the description of what it does), the reason it doesn't work on Staging is because we're explicitly ignoring zero length plays, at least for images. Would need testing on real hardware to know if this is correct or not.
> 
> https://github.com/dosbox-staging/dosbox-staging/blob/6c87b92dad66709a9f1c134a373472df8cd58217/src/dos/cdrom_image.cpp#L710-L718
> 
> > I assume that the chance of the seek implementation causing regressions with other titles that use CD Audio is very low. I also assume it works with OGG, MP3, FLAC, etc. audio tracks, but I haven't actually tested it.
> > I think it's safe to backport this to improve game compatibility. What's your opinion?
> 
> It's probably safe to backport. I can do a bit of regression testing. Is it documented anywhere how the .ogg files need to be structured? I never really use that feature but I seem to recall it needs a .cue file to be written out in a specific way?

You may also want to add `Chasm: The Rift` to your regressing testing since Staging 0.81 fixed a similar seeking issue.

https://dosbox-staging.github.io/releases/release-notes/0.81.0/

`CD-ROM seek delays are now emulated to fix the CD audio in a few games. For example, Chasm: The Rift stopped playing the music after about 20 minutes of gameplay without this fix.`

**Comment by weirddan455** [comment]
>  You may also want to add Chasm: The Rift to your regression testing since Staging 0.81 fixed a similar seeking issue.

Good call.  More about about that in the commit message here:  83fa5307f3e9362898662289ff6c41215ab1342c

I may need to insert a call to `LagDriveRespone()` if Chasm breaks.  I'll test later and report back.

**Comment by MasterO2** [comment]
> > You may also want to add Chasm: The Rift to your regression testing since Staging 0.81 fixed a similar seeking issue.
> 
> Good call. More about about that in the commit message here: [83fa530](https://github.com/dosbox-staging/dosbox-staging/commit/83fa5307f3e9362898662289ff6c41215ab1342c)
> 
> I may need to insert a call to `LagDriveRespone()` if Chasm breaks. I'll test later and report back.

I also recommend installing `Chasm: The Rift` via the instructions at https://github.com/dosbox-staging/dosbox-staging/wiki/Game-issues#chasm-the-rift.  The reason for that is if you try to run Chasm: The Rift with either `core = auto` or `core = dynamic` after the game is initially installed, Chasm's executable will display the error message `Runtime error 200 at 0002:1524`.

**Comment by johnnovak** [comment]
@weirddan455 You can convert BIN/CUE images to BIN/CUE + external WAV files using [this method](https://www.reddit.com/r/kde/comments/kmtxo6/comment/ghid8t8/?utm_source=share&utm_medium=mweb3x&utm_name=mweb3xcss&utm_term=1&utm_content=share_button). There are also lots of free Windows GUI tools for this, but I guess you prefer Linux CLI 😄 

Then it's a matter of converting the WAVs into OGG, MP3, FLAC, etc. and updating the file extensions in the CUE file that references them.

These contrib scripts might also help with testing:

https://github.com/dosbox-staging/dosbox-staging/blob/main/contrib/scripts/chapters-to-tracks.sh

https://github.com/dosbox-staging/dosbox-staging/blob/main/contrib/scripts/tracks-to-cues.sh

https://github.com/dosbox-staging/dosbox-staging/blob/main/contrib/scripts/playlist-to-tracks.sh

File: src/dos/dos_mscdex.cpp (hunk 1 above)
Comment (weirddan455) [reply src/dos/dos_mscdex.cpp:936]: Fixed

**Comment by weirddan455** [comment]
I had to write my own program to extract audio into .ogg files.  That `bchunk` program didn't like the way my .cue was formatted and threw an error.  The .ogg files play back in Dosbox though so no regressions there as far as I can tell.  Also, left Chasm running for about an hour so it would loop the track multiple times.  Audio kept playing the entire time.  Didn't run into any problems with that runtime error either.

I think this should be good to merge once CI passes.  Just pushed a formatting change.  Probably safe for backport if you want as well @johnnovak

**Comment by johnnovak** [comment]
Thanks for the testing @weirddan455. Yeah let's backport it as it's a low risk change and it improves game compatibility.

**Comment by MX9000** [comment]
@weirddan455 I tried the build changelog-v0.82.0-alpha-330-gd7ff.txt.zip and I found some issues.

Sound Blaster 16
- Silmarils logo animation: music OK
- Game intro with fighters presentation: music OK
- Game menu and screen with countdown before the fight: music audible but at very low volume
- Fight start: announcer voice OK but background music with very low volume

Sound Blaster
- Silmarils logo animation: music OK
- Game intro with fighters presentation: music missing and background buzz
- Game menu and screen with countdown before the fight: music OK
- Fight start: music OK but announcer voice and fight sounds with echo

Sound Blaster Pro
- Silmarils logo animation: music OK
- Game intro with fighters presentation: music OK
- Game menu and screen with countdown before the fight: music OK
- Fight start: announcer voice and music OK

**Comment by johnnovak** [comment]
> @weirddan455 I tried the build changelog-v0.82.0-alpha-330-gd7ff.txt.zip and I found some issues.
> 
> Sound Blaster 16
> 
>     * Silmarils logo animation: music OK
> 
>     * Game intro with fighters presentation: music OK
> 
>     * Game menu and screen with countdown before the fight: music audible but at very low volume
> 
>     * Fight start: announcer voice OK but background music with very low volume
> 
> 
> Sound Blaster
> 
>     * Silmarils logo animation: music OK
> 
>     * Game intro with fighters presentation: music missing and background buzz
> 
>     * Game menu and screen with countdown before the fight: music OK
> 
>     * Fight start: music OK but announcer voice and fight sounds with echo
> 
> 
> Sound Blaster Pro
> 
>     * Silmarils logo animation: music OK
> 
>     * Game intro with fighters presentation: music OK
> 
>     * Game menu and screen with countdown before the fight: music OK
> 
>     * Fight start: announcer voice and music OK

The game might as well behave exactly like this on real hardware too for all we know. This does not prove DOSBox does anything wrong.

**Comment by MX9000** [comment]
MIXER command did the trick!
Thank you!

**Pull request merged**